### ISSUE TITLE
`SocialAccount` not created when Django user found using the `SOCIALACCOUNT_EMAIL_AUTHENTICATION` workflow

### ISSUE THREAD

**Issue by agorka-contractsafe** [pennersr/django-allauth#3397]
First off, really appreciate you maintaining this library, it's very helpful. Thanks so much.

We've started implementing a SAML provider using email authentication, and we've noticed that when the email from the IdP matches one of our users' emails, a `SocialAccount` doesn't get created. But when a match isn't found a new user is provisioned just-in-time, it does. The user still gets logged in just fine in the existing user case, there's just no social account record for the user.

I wanted to check if this is the desired behavior. `SocialLogin.save` has the doc `Saves a new account. Note that while the account is new, the user may be an existing one (when connecting accounts)`, which makes me think we'd want to create a new social account when a user is connected for the first time. But I'm not sure not creating it is by design and something we should handle through something like the `social_account_updated` signal, or if it's a bug in the existing workflow.

Here are some details from some digging we've done, incase it's helpful:

We traced this back a bit and saw it starts in `socialaccount.helpers.complete_social_login` where the `sociallogin.lookup` occurs. In that function, once it fails the `_lookup_by_socialaccount` check and then moved to the email case, if there's a match, the user account gets set on `sociallogin.user`, but it won't save the social account to the DB. With the matching user now on the `sociallogin`, it then passes the `sociallogin.is_existing` check in `_complete_social_login`. From this point forward, there doesn't seem to be any logic to save social accounts.

We're on version 0.55.0.

Let me know if you have any questions or would like any clarification. Thank you!

**Comment by pennersr** [comment]
This is by design. The rationale is that when the user is signing in, a user would not expect his account to be modified as a result of that. Automatically adding new social accounts to a user account without the user knowingly initiating the action is unexpected behavior. The user can of course always explicitly choose to add the social account via the connection view.

Having said that, wondering about your use case though -- is it an issue in your case that the SAML account is not automatically added to the existing local user account, and if so, why?

**Comment by ryanhiebert** [comment]
Hi @pennersr, I work with @agorka-contractsafe integrating allauth and using it for SAML SSO.

I'm a little torn on it, but I'm leaning toward thinking that we should go ahead and make the social account. There are two reasons I'm thinking this.

First, I'm anticipating SCIM, where SAML or OpenID SSO may be matched via the IdP's ID to our internal ID. Because we don't get the Social Account in these cases, it means I won't have it to map it back correctly. SCIM can also be used to provision, and in those cases it won't matter, but if we're automatically creating accounts as well, it is important to have that mapping.

I can see an argument where if that data is important, then I shouldn't be relying on internal allauth models, particularly because this one gets mangled with the provider ID, and that's not without merit. The hooks provided should be sufficient to handle that use case. However, this seems like a potentially very useful and generic extension that we could make, perhaps in a third-party package, to allauth, and if the fields were intended to be relied on that way it could make it simpler of an implementation.

The second reason I have tended to think it might be best to have it create a social account is for consistency with automatically created accounts, where a social account is created as well. For any particular provider it seems like the login flow should either create social accounts or not, no matter whether it had to create a Django user account or not. If you agree with my consistency argument but not my SCIM argument, then you may wish to consider _not_ making a social account when you create a user account if it's based on email authentication.

If you do find yourself agreeing with the logic, but are still not sure whether you want to extend social account creation this way by default, you could create a setting.

-----

Separately, and now realizing that email authentication is a new feature that may have some growing pains with it, I'll note that there was a flow I was thinking about where it saw that the user exists already and they would be redirected to a login page to authenticate in some other way in order to validate that they wish to add it to the user account. It hadn't occurred to me that we might not create a social account instance at all, so that was a bit of a surprise to the way I was thinking about it.

For a little possibly useless context, we'd only want that behavior when the existing user was in a different tenant than the one that the provider is for, based on a separately maintained mapping, but allowing to to automatically create only for some tenant definitely would only make sense by using a hook or signal, and I think that the ones currently available would be sufficient for that case.

-----

Thank you for letting us get your feedback on this. It really helps us to understand your intended design for the library, so we can use it the right way and not be fighting your intentions with it. We really appreciate the quality of this library and the utility it provides us.

**Comment by pennersr** [comment]
Wondering about the SCIM use case. So, you get a SCIM request to create or update a user having ID `scim-123`. Given that this is not an issue for users that have a SAML account attached, I assume that you can look up those users using something of the form `SocialAccount.objects.get(uid="scim-123")`, and sync the SCIM data. Now, in case that no such `SocialAccount` exists, can't you simply do a `SocialAccount.objects.create(uid="scim-123")` ?

**Comment by ryanhiebert** [comment]
Yes, that's correct, I can use SCIM to create a SocialAccount that way, and that would be the right thing to do. What I won't be able to do is let SCIM correctly report the user accounts that _currently exist_ including that uid. So the provisioning part would work well, but not so much the querying. Even if they'd logged in via SSO before, it wouldn't have recorded the uid.

**Comment by ryanhiebert** [comment]
Nothing would be missing after a SCIM POST request for any particular user that was provisioned that way.

However, if a user was logged in via SAML before SCIM, such as with a JIT provisioning approach, and there was a matching account for that email for a different tenant, we wouldn't have a SocialAccount record to note that the uid was used to log in via the SAML provider, so we'd be missing the mapping from the user to the uid that SCIM would be looking at.

This affects GET requests that are querying the current state of the user database, or potentially DELETE requests that might be wanting to deprovision users.

It's also possible that I'm not deep enough into the weeds with SCIM and I'm missing something that would make this no big deal, so I apprecate you discussing it with me. Thank you.

We did also notice another odd way this manifests. The `social_account_updated` signal may run without any `SocialAccount` instance actually existing.

**Comment by ryanhiebert** [comment]
One more way that the existing behavior strikes me as inconsistent. The `uid` mapping for SAML is required, even when it won't be used to either match (because the uid doesn't exist yet) and it also won't be saved because its just matching on the uid.

Thanks again for your great work with this library, and allowing us to use that work. That I can only really find these kinds of subtle inconsistencies speaks volumes to me about how good the quality of the work you've done here is.

**Comment by pennersr** [comment]
Alright, details aside, there may indeed be a use case for having the social account automatically connected when you sign in via social account email authentication. In order to implement this, we need:

- A new setting: `SOCIALACCOUNT_EMAIL_AUTHENTICATION_AUTO_CONNECTS` (default: `False`) (or, is there a good shorter name?)

- It should trigger the `social_account_added` signal.

- TBD -- should it trigger the `socialaccount/messages/account_connected.txt` message? Likely not, as the user's mental model is that of signing in. Then again, that was the whole rationale for not connecting in the first place, and here we're doing it anyway, so do show the message?

PS: Thanks for the positive feedback...

**Comment by ryanhiebert** [comment]
Cool, that's great, thank you.

1. A settings works just fine for me. That name's long, but it works and I don't have a better suggestion.
2. What do you make of the consistency piece? Should automatically created accounts match pre-existing accounts using email authentication for whether or not a `SocialAccount` should be created?
3. I _think_ that `social_account_added` is currently called, even in this circumstance where no social account is actually created, should it *not* be called when the social account isn't actually added because this setting is `False`? Or is this inconsistency acceptable, you think? If I were you, I'd be pretty torn, because if I didn't like the inconsistency I'd probably think I'd need to add a different signal to handle the case where there was not a matching `SocialAccount`.
4. I'm not sure on whether to show that template. I think for my uninformed hypothetical use case I would typically not want to bother with showing the connection, because I'm wanting to make the connection/login just magically work. Maybe, if I were designing it, *this* is what I'd create some kind of setting for, and not bother with the `AUTO_CONNECTS` setting and just always make the `SocialAccount`. It would make all the inconsistency questions I've raised irrelevant.

**Comment by pennersr** [comment]
With respect to 3, I cannot confirm that this happens for logins using `SOCIALACCOUNT_EMAIL_AUTHENTICATION`. Looking at the code:

- All social logins go through `def complete_social_login(request, sociallogin):`
- If, and only if, the process is `AuthProcess.CONNECT` the method `_add_social_account()` is called. That method is the only place where `signals.social_account_updated.send()` is invoked.

So I don't see how we could end up sending that signal?

**Comment by joergmschulz** [comment]
From process point of view: assuming email addresses are unique, the best flow would be (for users who logged in conventionally earlier): 
- user logs in via social
- if there is already a user with the {email} in scope, but no social user yet, the user link is created automatically.
- if there is no such user, the user is created completely just like today.
INMHO this should be the default behaviour, but be configurable as you suggested above. To be checked: what happens when a user logs in via multiple social sources (like github, OIDC, SAML)? Create a link for every one of these sources?

**Comment by ryanhiebert** [comment]
@pennersr sorry for the delay, I'm going to check on your comment today or tomorrow with @agorka-contractsafe .

@joergmschulz I'd expected that we'd get a SocialAccount for each source, yes, and while its my preference as well that it would be the default (and instead have a setting to determine whether or not to show the account_connected template, as I described a little earlier), this is @pennersr package and I'm happy to use his package the way he chooses to design it. I can definitely live with a setting that gives me the mode I am looking for.

**Comment by pennersr** [comment]
Closed via ab704687

**Comment by agorka-contractsafe** [comment]
> With respect to 3, I cannot confirm that this happens for logins using SOCIALACCOUNT_EMAIL_AUTHENTICATION. Looking at the code:
> - All social logins go through def complete_social_login(request, sociallogin):
> - If, and only if, the process is AuthProcess.CONNECT the method _add_social_account() is called. That method is the only place where signals.social_account_updated.send() is invoked.
>
> So I don't see how we could end up sending that signal?

Hey @pennersr. Sorry, should clarify the signal we meant was `social_account_updated`, which is called in `_complete_social_login`:

```python
if sociallogin.is_existing:
    # Login existing user
    ret = _login_social_account(request, sociallogin)
    signals.social_account_updated.send(
        sender=SocialLogin, request=request, sociallogin=sociallogin
    )
```

In this case, no social account is updated, since we can get here without a social account existing, which could be misleading. It is also called in the `_add_social_account` method, which you mention as well.

**Comment by pennersr** [comment]
That is indeed an issue -- incorrectly sending `social_account_updated` is addressed via 51e64afb

**Comment by pennersr** [comment]
https://codeberg.org/allauth/django-allauth/issues/3397
